oxidation/libparsec/crates/libparsec_fs/src/file_operations.rs

```diff
@@ -0,0 +1,358 @@
+    size: u64,
+    start: u64,
+    new_chunk: Chunk,
+) -> (Vec<Chunk>, HashSet<ChunkID>) {
+    let stop = start + size;
+
+    // Edge case
+    if chunks.is_empty() {
+        return (vec![new_chunk], HashSet::new());
+    }
+
+    // Bisect
+    let start_index = match chunks.binary_search_by_key(&start, |x| x.start) {
+        Ok(x) => x,
+        Err(x) => x
+            .checked_sub(1)
+            .expect("First chunk should always start at 0"),
+    };
+    let stop_index = match chunks.binary_search_by_key(&stop, |x| x.start) {
+        Ok(x) => x,
+        Err(x) => x,
+    };
+
+    // Removed ids
+    let mut removed_ids: HashSet<ChunkID> = chunks
+        .get(start_index..stop_index)
+        .unwrap_or_default()
+        .iter()
+        .map(|x| x.id)
+        .collect();
+
+    // New chunks
+    let mut new_chunks: Vec<Chunk> = chunks.get(0..start_index).unwrap_or_default().to_vec();
+
+    // Test start chunk
+    let start_chunk = chunks
+        .get(start_index)
+        .expect("Indexes are found using binary search and hence always valid");
+    if start_chunk.start < start {
+        let mut new_start_chunk = start_chunk.clone();
+        new_start_chunk.stop = NonZeroU64::new(start)
+            .expect("Cannot be zero since it's strictly greater than start_chunk.start");
+        new_chunks.push(new_start_chunk);
+        removed_ids.remove(&start_chunk.id);
+    }
+
+    // Add new buffer
+    new_chunks.push(new_chunk);
+
+    // Test stop_chunk
+    let stop_chunk = chunks
+        .get(stop_index - 1)
+        .expect("Indexes are found using binary search and hence always valid");
+    if stop_chunk.stop.get() > stop {
+        let mut new_stop_chunk = stop_chunk.clone();
+        new_stop_chunk.start = stop;
+        new_chunks.push(new_stop_chunk);
+        removed_ids.remove(&stop_chunk.id);
+    }
+
+    // Fill up
+    new_chunks.extend(chunks.get(stop_index..).unwrap_or_default().iter().cloned());
+
+    // IDs might appear multiple times
+    for chunk in &new_chunks {
+        removed_ids.remove(&chunk.id);
+        if removed_ids.is_empty() {
+            break;
+        }
+    }
+
+    (new_chunks, removed_ids)
+}
+
+pub fn prepare_write(
+    manifest: &LocalFileManifest,
+    mut size: u64,
+    mut offset: u64,
+    timestamp: DateTime,
+) -> (LocalFileManifest, Vec<WriteOperation>, HashSet<ChunkID>) {
+    let mut padding = 0;
+    let mut removed_ids = HashSet::new();
+    let mut write_operations = vec![];
+
+    // Padding
+    if offset > manifest.size {
+        padding = offset - manifest.size;
+        size += padding;
+        offset = manifest.size;
+    }
+
+    // Copy buffers
+    let mut blocks = manifest.blocks.clone();
+
+    // Find proper block indexes
+    let blocksize = u64::from(manifest.blocksize);
+    let start_block = offset / blocksize;
+    let stop_block = (offset + size + blocksize - 1) / blocksize;
+
+    // Loop over blocks
+    for block in start_block..stop_block {
```

Comment:
Maybe we should cast start_block as usize and stop_block as usize once
then we can use block serenely

Reply:
Casting `block` to `usize` made sense in `prepare_truncate` but here it's more annoying that helping since there are other conflicting operations using `u64`.

Reply:
Nvm if it's annoying